pkg/resources/digitalocean/dns/dns.go

```diff
@@ -0,0 +1,501 @@
+func (z *zones) List() ([]dnsprovider.Zone, error) {
+	domains, err := listDomains(z.client)
+	if err != nil {
+		return nil, err
+	}
+
+	var newZone *zone
+	var zones []dnsprovider.Zone
+	for _, domain := range domains {
+		newZone = &zone{
+			name:   domain.Name,
+			client: z.client,
+		}
+		zones = append(zones, newZone)
+	}
+
+	return zones, nil
+}
+
+// Add adds a new DNS zone
+func (z *zones) Add(newZone dnsprovider.Zone) (dnsprovider.Zone, error) {
+	domainCreateRequest := &godo.DomainCreateRequest{
+		Name:      newZone.Name(),
+		IPAddress: ipPlaceholder,
+	}
+
+	domain, err := createDomain(z.client, domainCreateRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return &zone{
+		name:   domain.Name,
+		client: z.client,
+	}, nil
+}
+
+// Remove deletes a zone
+func (z *zones) Remove(zone dnsprovider.Zone) error {
+	return deleteDomain(z.client, zone.Name())
+}
+
+// New returns a new implementation of dnsprovider.Zone
+func (z *zones) New(name string) (dnsprovider.Zone, error) {
+	return &zone{
+		name:   name,
+		client: z.client,
+	}, nil
+
+}
+
+// zone implements dnsprovider.Zone
+type zone struct {
+	name   string
+	client *godo.Client
+}
+
+// Name returns the Name of a dns zone
+func (z *zone) Name() string {
+	return z.name
+}
+
+// ID returns the name of a dns zone, in DO the ID is the name
+func (z *zone) ID() string {
+	return z.name
+}
+
+// ResourceRecordSet returns an implementation of dnsprovider.ResourceRecordSets
+func (z *zone) ResourceRecordSets() (dnsprovider.ResourceRecordSets, bool) {
+	return &resourceRecordSets{zone: z, client: z.client}, true
+}
+
+// resourceRecordSets implements dnsprovider.ResourceRecordSet
+type resourceRecordSets struct {
+	zone   *zone
+	client *godo.Client
+}
+
+// List returns a list of dnsprovider.ResourceRecordSet
+func (r *resourceRecordSets) List() ([]dnsprovider.ResourceRecordSet, error) {
+	records, err := getRecords(r.client, r.zone.Name())
+	if err != nil {
+		return nil, err
+	}
+
+	var rrset *resourceRecordSet
+	var rrsets []dnsprovider.ResourceRecordSet
+	for _, record := range records {
+		rrset = &resourceRecordSet{
+			name:       record.Name,
+			data:       record.Data,
+			ttl:        record.TTL,
+			recordType: rrstype.RrsType(record.Type),
+		}
+
+		rrsets = append(rrsets, rrset)
+	}
+
+	return rrsets, nil
+
+}
+
+// Get returns a list of dnsprovider.ResourceRecordSet that matches the name parameter
+func (r *resourceRecordSets) Get(name string) ([]dnsprovider.ResourceRecordSet, error) {
+	records, err := r.List()
+	if err != nil {
+		return nil, err
+	}
+
+	var recordSets []dnsprovider.ResourceRecordSet
+	for _, record := range records {
+		if record.Name() == name {
+			recordSets = append(recordSets, record)
+		}
+	}
+
+	return recordSets, nil
+}
+
+// New returns an implementation of dnsprovider.ResourceRecordSet
+func (r *resourceRecordSets) New(name string, rrdatas []string, ttl int64, rrstype rrstype.RrsType) dnsprovider.ResourceRecordSet {
```

Comment:
I believe all the values are required, I would return an err in this function, but I'm using the dnsprovider interface https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/dns.go#L63